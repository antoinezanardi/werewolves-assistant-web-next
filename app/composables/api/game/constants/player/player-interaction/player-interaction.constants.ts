const PLAYER_INTERACTION_TYPES = [
  "eat",
  "look",
  "charm",
  "give-life-potion",
  "give-death-potion",
  "shoot",
  "protect",
  "mark",
  "vote",
  "choose-as-model",
  "choose-as-sheriff",
  "sniff",
  "ban-voting",
  "transfer-sheriff-role",
  "sentence-to-death",
  "steal-role",
  "infect",
  "bury",
] as const;

export { PLAYER_INTERACTION_TYPES };