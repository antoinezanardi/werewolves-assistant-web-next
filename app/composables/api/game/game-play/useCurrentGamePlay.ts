import type { ComputedRef, Ref } from "vue";
import type { GamePlayCause } from "~/composables/api/game/types/game-play/game-play.types";

import type { Game } from "~/composables/api/game/types/game.class";
import type { PlayerInteractionType } from "~/composables/api/game/types/players/player-interaction/player-interaction.types";
import type { Player } from "~/composables/api/game/types/players/player.class";

type UseCurrentGamePlay = {
  mustCurrentGamePlayBeSkipped: ComputedRef<boolean>;
  priorityCauseInCurrentGamePlay: ComputedRef<GamePlayCause | undefined>;
  getEligibleTargetsWithInteractionInCurrentGamePlay: (interaction: PlayerInteractionType) => Player[];
};

function useCurrentGamePlay(game: Ref<Game>): UseCurrentGamePlay {
  const mustCurrentGamePlayBeSkipped = computed<boolean>(() => {
    const isWolfHoundSideRandomlyChosen = game.value.options.roles.wolfHound.isSideRandomlyChosen;
    const stealRoleEligibleTargets = getEligibleTargetsWithInteractionInCurrentGamePlay("steal-role");
    const currentGameAction = game.value.currentPlay?.action;

    return currentGameAction === "bury-dead-bodies" && !stealRoleEligibleTargets.length ||
      currentGameAction === "choose-side" && isWolfHoundSideRandomlyChosen;
  });

  const priorityCauseInCurrentGamePlay = computed<GamePlayCause | undefined>(() => {
    const gamePlayCausesSortedByPriority: GamePlayCause[] = [
      "previous-votes-were-in-ties",
      "angel-presence",
      "stuttering-judge-request",
    ];

    return gamePlayCausesSortedByPriority.find(cause => game.value.currentPlay?.causes?.includes(cause));
  });

  function getEligibleTargetsWithInteractionInCurrentGamePlay(interactionType: PlayerInteractionType): Player[] {
    if (game.value.currentPlay?.source.interactions === undefined) {
      return [];
    }
    const interaction = game.value.currentPlay.source.interactions.find(({ type }) => type === interactionType);
    if (interaction === undefined) {
      return [];
    }
    return interaction.eligibleTargets;
  }
  return {
    mustCurrentGamePlayBeSkipped,
    priorityCauseInCurrentGamePlay,
    getEligibleTargetsWithInteractionInCurrentGamePlay,
  };
}

export { useCurrentGamePlay };