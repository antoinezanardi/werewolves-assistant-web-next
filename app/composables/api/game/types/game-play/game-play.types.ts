import type { TupleToUnion } from "type-fest";

import type { GAME_PLAY_TYPES, GAME_PLAY_CAUSES, GAME_PLAY_OCCURRENCES, WITCH_POTIONS, GAME_PLAY_ACTIONS } from "~/composables/api/game/constants/game-play/game-play.constants";

type GamePlayType = TupleToUnion<typeof GAME_PLAY_TYPES>;

type GamePlayAction = TupleToUnion<typeof GAME_PLAY_ACTIONS>;

type GamePlayCause = TupleToUnion<typeof GAME_PLAY_CAUSES>;

type GamePlayOccurrence = TupleToUnion<typeof GAME_PLAY_OCCURRENCES>;

type WitchPotion = TupleToUnion<typeof WITCH_POTIONS>;

export type {
  GamePlayType,
  GamePlayAction,
  GamePlayCause,
  GamePlayOccurrence,
  WitchPotion,
};