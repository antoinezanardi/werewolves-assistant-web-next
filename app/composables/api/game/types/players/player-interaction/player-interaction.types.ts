import type { TupleToUnion } from "type-fest";

import type { PLAYER_INTERACTION_TYPES } from "~/composables/api/game/constants/player/player-interaction/player-interaction.constants";

type PlayerInteractionType = TupleToUnion<typeof PLAYER_INTERACTION_TYPES>;

export type { PlayerInteractionType };