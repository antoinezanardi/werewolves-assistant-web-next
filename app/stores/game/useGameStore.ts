import type { AsyncDataRequestStatus } from "nuxt/app";
import { defineStore } from "pinia";
import { unique } from "radash";

import type { MakeGamePlayDto } from "~/composables/api/game/dto/make-game-play/make-game-play.dto";
import type { GameOptions } from "~/composables/api/game/types/game-options/game-options.class";
import { Game } from "~/composables/api/game/types/game.class";
import { useFetchGames } from "~/composables/api/game/useFetchGames";
import { StoreIds } from "~/stores/enums/store.enum";
import { useGameEventsStore } from "~/stores/game/game-event/useGameEventsStore";

const useGameStore = defineStore(StoreIds.GAME, () => {
  const {
    getGame: fetchGameFromApi,
    cancelGame: cancelGameFromApi,
    makeGamePlay: makeGamePlayFromApi,
  } = useFetchGames();

  const game = ref<Game>(new Game());
  const gameOptions = computed<GameOptions>(() => game.value.options);

  const gamePlayerGroups = computed<string[]>(() => unique(game.value.players.map(player => player.group)).filter(Boolean) as string[]);

  const fetchingGameStatus = ref<AsyncDataRequestStatus>("idle");
  const cancelingGameStatus = ref<AsyncDataRequestStatus>("idle");
  const makingGamePlayStatus = ref<AsyncDataRequestStatus>("idle");

  const { resetGameEventIndex } = useGameEventsStore();

  function resetGame(): void {
    game.value = new Game();
  }

  async function fetchAndSetGame(gameId: string): Promise<void> {
    fetchingGameStatus.value = "pending";
    const fetchedGame = await fetchGameFromApi(gameId);
    if (!fetchedGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    resetGameEventIndex();
    game.value = fetchedGame;
    fetchingGameStatus.value = "success";
  }

  async function cancelGame(): Promise<void> {
    cancelingGameStatus.value = "pending";
    const canceledGame = await cancelGameFromApi(game.value._id);
    if (!canceledGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    game.value = canceledGame;
    cancelingGameStatus.value = "success";
  }

  async function makeGamePlay(makeGamePlayDto: MakeGamePlayDto): Promise<void> {
    makingGamePlayStatus.value = "pending";
    const playedGame = await makeGamePlayFromApi(game.value._id, makeGamePlayDto);
    if (!playedGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    resetGameEventIndex();
    game.value = playedGame;
    makingGamePlayStatus.value = "success";
  }

  async function skipGamePlay(): Promise<void> {
    return makeGamePlay({});
  }
  return {
    game,
    gameOptions,
    gamePlayerGroups,
    fetchingGameStatus,
    cancelingGameStatus,
    makingGamePlayStatus,
    resetGame,
    fetchAndSetGame,
    cancelGame,
    makeGamePlay,
    skipGamePlay,
  };
});

export { useGameStore };