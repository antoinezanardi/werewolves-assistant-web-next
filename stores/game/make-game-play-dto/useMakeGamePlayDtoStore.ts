import { defineStore } from "pinia";

import { MakeGamePlayTargetDto } from "~/composables/api/game/dto/make-game-play/make-game-play-target/make-game-play-target.dto";
import { MakeGamePlayVoteDto } from "~/composables/api/game/dto/make-game-play/make-game-play-vote/make-game-play-vote.dto";
import { MakeGamePlayDto } from "~/composables/api/game/dto/make-game-play/make-game-play.dto";
import type { RoleSides } from "~/composables/api/role/enums/role.enums";
import { StoreIds } from "~/stores/enums/store.enum";

const useMakeGamePlayDtoStore = defineStore(StoreIds.MAKE_GAME_PLAY_DTO, () => {
  const makeGamePlayDto = ref<MakeGamePlayDto>(MakeGamePlayDto.create({}));

  function setMakeGamePlayDto(makeGamePlayDtoValue: MakeGamePlayDto): void {
    makeGamePlayDto.value = MakeGamePlayDto.create(makeGamePlayDtoValue);
  }

  function resetMakeGamePlayDto(): void {
    makeGamePlayDto.value = MakeGamePlayDto.create({});
  }

  function addMakeGamePlayTargetDto(target: MakeGamePlayTargetDto): void {
    if (!makeGamePlayDto.value.targets) {
      makeGamePlayDto.value.targets = [];
    }
    makeGamePlayDto.value.targets.push(MakeGamePlayTargetDto.create(target));
  }

  function removeMakeGamePlayTargetDto(targetId: string): void {
    if (!makeGamePlayDto.value.targets) {
      return;
    }
    const targetIndex = makeGamePlayDto.value.targets.findIndex(target => target.playerId === targetId);
    if (targetIndex !== -1) {
      makeGamePlayDto.value.targets = makeGamePlayDto.value.targets.toSpliced(targetIndex, 1);
    }
    if (makeGamePlayDto.value.targets.length === 0) {
      makeGamePlayDto.value.targets = undefined;
    }
  }

  function addMakeGamePlayVoteDto(vote: MakeGamePlayVoteDto): void {
    if (!makeGamePlayDto.value.votes) {
      makeGamePlayDto.value.votes = [];
    }
    makeGamePlayDto.value.votes.push(MakeGamePlayVoteDto.create(vote));
  }

  function removeMakeGamePlayVoteDto(sourceId: string): void {
    if (!makeGamePlayDto.value.votes) {
      return;
    }
    const voteIndex = makeGamePlayDto.value.votes.findIndex(vote => vote.sourceId === sourceId);
    if (voteIndex !== -1) {
      makeGamePlayDto.value.votes = makeGamePlayDto.value.votes.toSpliced(voteIndex, 1);
    }
    if (makeGamePlayDto.value.votes.length === 0) {
      makeGamePlayDto.value.votes = undefined;
    }
  }

  function setDoesJudgeRequestAnotherVote(doesJudgeRequestAnotherVote: boolean | undefined): void {
    makeGamePlayDto.value.doesJudgeRequestAnotherVote = doesJudgeRequestAnotherVote;
  }

  function setChosenCardId(chosenCardId: string | undefined): void {
    makeGamePlayDto.value.chosenCardId = chosenCardId;
  }

  function setChosenSide(chosenSide: RoleSides | undefined): void {
    makeGamePlayDto.value.chosenSide = chosenSide;
  }
  return {
    makeGamePlayDto,
    setMakeGamePlayDto,
    resetMakeGamePlayDto,
    addMakeGamePlayTargetDto,
    removeMakeGamePlayTargetDto,
    addMakeGamePlayVoteDto,
    removeMakeGamePlayVoteDto,
    setDoesJudgeRequestAnotherVote,
    setChosenCardId,
    setChosenSide,
  };
});

export { useMakeGamePlayDtoStore };