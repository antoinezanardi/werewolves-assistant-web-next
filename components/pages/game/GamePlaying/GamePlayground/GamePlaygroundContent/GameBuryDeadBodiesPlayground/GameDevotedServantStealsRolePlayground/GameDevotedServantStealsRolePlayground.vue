<template>
  <div
    id="game-devoted-servant-steal-role-playground"
    class="flex flex-col items-center"
  >
    <GlowCapture>
      <GlowElement>
        <RoleImage
          class="glow:border-gray-400"
          definition="normal"
          role-name="devoted-servant"
          sizes="175"
        />
      </GlowElement>
    </GlowCapture>

    <h3
      id="devoted-servant-steals-role-question"
      class="my-4"
    >
      {{ doesDevotedServantStealRoleQuestion }}
    </h3>

    <div
      id="targets"
      class="flex items-center justify-center"
    >
      <GamePlaygroundPlayerCard
        v-for="target in eliminatedPlayers"
        :key="target._id"
        class="target"
        interaction="steal-role"
        :player="target"
      />
    </div>
  </div>
</template>

<script setup lang="ts">
import { storeToRefs } from "pinia";

import GamePlaygroundPlayerCard from "~/components/pages/game/GamePlaying/GamePlayground/GamePlaygroundContent/GamePlaygroundPlayerCard/GamePlaygroundPlayerCard.vue";
import RoleImage from "~/components/shared/role/RoleImage/RoleImage.vue";
import { useCurrentGamePlay } from "~/composables/api/game/game-play/useCurrentGamePlay";
import type { Player } from "~/composables/api/game/types/players/player.class";
import { useGameStore } from "~/stores/game/useGameStore";

const gameStore = useGameStore();
const { game } = storeToRefs(gameStore);

const { t } = useI18n();

const { getEligibleTargetsWithInteractionInCurrentGamePlay } = useCurrentGamePlay(game);

const eliminatedPlayers = computed<Player[] | undefined>(() => getEligibleTargetsWithInteractionInCurrentGamePlay("steal-role"));

const doesDevotedServantStealRoleQuestion = computed<string>(() => {
  const eliminatedPlayersCount = eliminatedPlayers.value?.length ?? 0;

  return t("components.GameDevotedServantStealsRolePlayground.doesDevotedServantStealRole", eliminatedPlayersCount);
});
</script>