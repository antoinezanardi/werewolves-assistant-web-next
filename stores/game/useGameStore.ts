import type { AsyncDataRequestStatus } from "#app/composables/asyncData";
import { defineStore } from "pinia";

import { Game } from "~/composables/api/game/types/game.class";
import { useFetchGames } from "~/composables/api/game/useFetchGames";
import { StoreIds } from "~/stores/enums/store.enum";

const useGameStore = defineStore(StoreIds.GAME, () => {
  const { getGame: fetchGameFromApi } = useFetchGames();

  const game = ref<Game>(new Game());
  const fetchingGameStatus = ref<AsyncDataRequestStatus>("idle");

  async function fetchAndSetGame(gameId: string): Promise<void> {
    fetchingGameStatus.value = "pending";
    const fetchedGame = await fetchGameFromApi(gameId);
    if (!fetchedGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    game.value = fetchedGame;
    fetchingGameStatus.value = "success";
  }
  return {
    game,
    fetchingGameStatus,
    fetchAndSetGame,
  };
});

export { useGameStore };