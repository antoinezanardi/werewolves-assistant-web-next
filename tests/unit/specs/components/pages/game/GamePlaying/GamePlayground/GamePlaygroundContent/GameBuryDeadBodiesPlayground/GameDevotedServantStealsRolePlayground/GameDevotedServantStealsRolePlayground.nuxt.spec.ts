import { createTestingPinia } from "@pinia/testing";
import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";
import { expect } from "vitest";

import type { VuePrimeToggleButton } from "#components";
import GameDevotedServantStealsRolePlayground from "~/components/pages/game/GamePlaying/GamePlayground/GamePlaygroundContent/GameBuryDeadBodiesPlayground/GameDevotedServantStealsRolePlayground/GameDevotedServantStealsRolePlayground.vue";
import { RoleNames } from "~/composables/api/role/enums/role.enums";
import { StoreIds } from "~/stores/enums/store.enum";
import { useMakeGamePlayDtoStore } from "~/stores/game/make-game-play-dto/useMakeGamePlayDtoStore";
import { useGameStore } from "~/stores/game/useGameStore";
import { createFakeMakeGamePlayTargetDto } from "~/tests/unit/utils/factories/composables/api/game/dto/make-game-play/make-game-play-target/make-game-play-target.dto.factory";
import { createFakeGamePlayEligibleTargets } from "~/tests/unit/utils/factories/composables/api/game/game-play/game-play-eligible-targets/game-play-eligible-targets.factory";
import { createFakeInteractablePlayer } from "~/tests/unit/utils/factories/composables/api/game/game-play/game-play-eligible-targets/interactable-player/interactable-player.factory";
import { createFakePlayerInteraction } from "~/tests/unit/utils/factories/composables/api/game/game-play/game-play-eligible-targets/interactable-player/player-interaction/player-interaction.factory";
import { createFakeGamePlaySurvivorsBuryDeadBodies } from "~/tests/unit/utils/factories/composables/api/game/game-play/game-play.factory";
import { createFakeGame } from "~/tests/unit/utils/factories/composables/api/game/game.factory";
import { createFakeSeerAlivePlayer } from "~/tests/unit/utils/factories/composables/api/game/player/player-with-role.factory";
import { mountSuspendedComponent } from "~/tests/unit/utils/helpers/mount.helpers";

describe("Game Devoted Servant Steals Role Playground Component", () => {
  let wrapper: ReturnType<typeof mount<typeof GameDevotedServantStealsRolePlayground>>;
  const eliminatePlayer = createFakeSeerAlivePlayer({ name: "Antoine" });
  const testingPinia = {
    initialState: {
      [StoreIds.GAME]: {
        game: createFakeGame({
          currentPlay: createFakeGamePlaySurvivorsBuryDeadBodies({
            eligibleTargets: createFakeGamePlayEligibleTargets({
              interactablePlayers: [
                createFakeInteractablePlayer({
                  player: eliminatePlayer,
                  interactions: [
                    createFakePlayerInteraction({
                      source: RoleNames.DEVOTED_SERVANT,
                      type: "steal-role",
                    }),
                  ],
                }),
              ],
            }),
          }),
        }),
      },
    },
  };

  async function mountGameDevotedServantStealsRolePlaygroundComponent(options: ComponentMountingOptions<typeof GameDevotedServantStealsRolePlayground> = {}):
  Promise<ReturnType<typeof mount<typeof GameDevotedServantStealsRolePlayground>>> {
    return mountSuspendedComponent(GameDevotedServantStealsRolePlayground, {
      global: { plugins: [createTestingPinia(testingPinia)] },
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountGameDevotedServantStealsRolePlaygroundComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Devoted Servant steals role question", () => {
    it("should translate devoted servant steals role question when player is defined.", () => {
      const devotedServantStealsRoleQuestion = wrapper.find<HTMLHeadingElement>("#devoted-servant-steals-role-question");

      expect(devotedServantStealsRoleQuestion.text()).toBe("components.GameDevotedServantStealsRolePlayground.doesDevotedServantStealRole, {\"playerName\":\"Antoine\"}");
    });

    it("should translate devoted servant steals role question when player is undefined.", async() => {
      const gameStore = useGameStore();
      gameStore.game.currentPlay = null;
      const devotedServantStealsRoleQuestion = wrapper.find<HTMLHeadingElement>("#devoted-servant-steals-role-question");
      await nextTick();

      expect(devotedServantStealsRoleQuestion.text()).toBe("components.GameDevotedServantStealsRolePlayground.doesDevotedServantStealRole, {}");
    });
  });

  describe("Toggle Button", () => {
    it("should translate off button label when rendered.", async() => {
      wrapper = await mountGameDevotedServantStealsRolePlaygroundComponent({ shallow: false });
      const toggleButton = wrapper.findComponent<typeof VuePrimeToggleButton>("#does-devoted-servant-steal-role-button");

      expect(toggleButton.props("offLabel")).toBe("She doesn't steal the role");
    });

    it("should translate on button label when rendered.", async() => {
      wrapper = await mountGameDevotedServantStealsRolePlaygroundComponent({ shallow: false });
      const toggleButton = wrapper.findComponent<typeof VuePrimeToggleButton>("#does-devoted-servant-steal-role-button");

      expect(toggleButton.props("onLabel")).toBe("She steals the role");
    });

    it("should set the eliminated player as target when toggle emits true change event.", async() => {
      wrapper = await mountGameDevotedServantStealsRolePlaygroundComponent({ shallow: false });
      const makeGamePlayDtoStore = useMakeGamePlayDtoStore();
      const toggleButtonCheckbox = wrapper.find("#does-devoted-servant-steal-role-button > .p-togglebutton-input");
      await toggleButtonCheckbox.setValue(true);
      const expectedMakeGamePlayTargetDto = createFakeMakeGamePlayTargetDto({ playerId: eliminatePlayer._id });

      expect(makeGamePlayDtoStore.addMakeGamePlayTargetDto).toHaveBeenCalledExactlyOnceWith(expectedMakeGamePlayTargetDto);
      expect(makeGamePlayDtoStore.removeMakeGamePlayTargetDto).not.toHaveBeenCalled();
    });

    it("should remove the eliminated player as target when toggle emits first true and false change event.", async() => {
      wrapper = await mountGameDevotedServantStealsRolePlaygroundComponent({ shallow: false });
      const makeGamePlayDtoStore = useMakeGamePlayDtoStore();
      const toggleButtonCheckbox = wrapper.find("#does-devoted-servant-steal-role-button > .p-togglebutton-input");
      await toggleButtonCheckbox.setValue(true);
      await toggleButtonCheckbox.setValue(false);

      expect(makeGamePlayDtoStore.removeMakeGamePlayTargetDto).toHaveBeenCalledExactlyOnceWith(eliminatePlayer._id);
    });

    it("should throw error when toggle emits change events but eliminated player is not set.", async() => {
      wrapper = await mountGameDevotedServantStealsRolePlaygroundComponent({ shallow: false });
      const gameStore = useGameStore();
      gameStore.game.currentPlay = null;
      const makeGamePlayDtoStore = useMakeGamePlayDtoStore();
      await nextTick();
      const toggleButtonCheckbox = wrapper.find("#does-devoted-servant-steal-role-button > .p-togglebutton-input");
      await toggleButtonCheckbox.setValue(true);

      expect(createError).toHaveBeenCalledExactlyOnceWith("Eliminated player is not found.");
      expect(makeGamePlayDtoStore.addMakeGamePlayTargetDto).not.toHaveBeenCalled();
      expect(makeGamePlayDtoStore.removeMakeGamePlayTargetDto).not.toHaveBeenCalled();
    });
  });
});