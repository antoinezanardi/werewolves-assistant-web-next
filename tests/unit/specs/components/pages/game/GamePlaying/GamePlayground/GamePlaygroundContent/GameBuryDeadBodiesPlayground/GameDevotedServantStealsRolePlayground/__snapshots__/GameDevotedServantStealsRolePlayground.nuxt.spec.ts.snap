// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Devoted Servant Steals Role Playground Component > should match snapshot when rendered. 1`] = `
"<div id="game-devoted-servant-steal-role-playground" class="flex flex-col items-center">
  <anonymous-stub class="" size="400" rest="[object Object]"></anonymous-stub>
  <h3 id="devoted-servant-steals-role-question" class="my-4">components.GameDevotedServantStealsRolePlayground.doesDevotedServantStealRole, 2</h3>
  <div id="targets" class="flex items-center justify-center">
    <game-playground-player-card-stub interaction="steal-role" player="[object Object]" class="target"></game-playground-player-card-stub>
    <game-playground-player-card-stub interaction="steal-role" player="[object Object]" class="target"></game-playground-player-card-stub>
  </div>
</div>"
`;
